### ISSUE TITLE
Kernel: Some kind of KASLR

### ISSUE THREAD

**Issue by N00byEdge** [SerenityOS/serenity#1039]
I've seen this mentioned in a few videos but I didn't find an issue on it, and I thought it might be good to make one to track the progress here. Making the layout less predictable would not help with making the kernel more secure, but harder to exploit if such a vulnerability is found, so this probably not of any real importance until people actually start using this as their everyday operating systems and it might be targeted in such a way.

One of the harder problems is finding an entropy source that early in the boot. Maybe there could be support for RDRAND when the hardware supports it? (This is testable in qemu with `-kvm -cpu host` if your host has it and requires setting no bits to enable)

I think a rough outline of things one would like to do is something like this:

- [ ] Establish partial entropy source
- [ ] Randomize kernel base
- [ ] Randomize kernel dynamic allocators
- [ ] Randomize userspace layout
- [ ] Provide randomness for userspace(?)
- [ ] Establish good entropy source

I guess a 64 bit switch also could be considered. 64 bits is not so relevant for the extra virtual address space (the important part is having any randomization at all, not lots), but rather for the pc-relative addressing, but that's not that big of a deal as your C++ compiler will insert a `get_pc_thunk` into your code here and there.

**Comment by davidchisnall** [comment]
These are, unfortunately, the easy bits of KASL.  The difficult bit is ensuring that you do not leak kernel addresses to userspace (otherwise you've added complexity for no security gain).  This is really, really hard to do if your kernel interfaces are not designed from the ground up to provide that guarantee.  It's proving very hard to retrofit to Linux and I think it would be impossible for Windows.  

Things are somewhat different if you do more than randomise the base.  For example, OpenBSD can re-link the kernel objects in a random order at boot time.  This makes it harder to find a specific function in the kernel, but does not help much from data-oriented attacks.

**Comment by bcoles** [comment]
Step 2 `Randomize kernel base`  was implemented in #13179.

**Comment by IdanHo** [comment]
I believe `Randomize userspace layout`, `Provide randomness for userspace(?)` and `Establish good entropy source` are also implemented. We randomize the placement of libraries/code/data/etc in process address spaces, we provide good randomness for userspace via `sys$getrandom`, and we fetch entropy from RDRAND/RDSEED/HPET/RTC/Interrupts/VirtIORand/etc.

**Comment by bcoles** [comment]
> I believe `Randomize userspace layout`, `Provide randomness for userspace(?)` and `Establish good entropy source` are also implemented. We randomize the placement of libraries/code/data/etc in process address spaces, we provide good randomness for userspace via `sys$getrandom`, and we fetch entropy from RDRAND/RDSEED/HPET/RTC/Interrupts/VirtIORand/etc.

Makes sense. This issue pre-dates introduction of userspace ASLR (#1048) and the numerous new randomization sources.

**Comment by IdanHo** [comment]
Indeed, making `kmalloc` and MM's allocate_kernel_region family of methods return randomised addresses is the next big task
